File: utils/benchmark/LinkedList/link.m
#import <Foundation/Foundation.h>
#include <mach/mach_time.h>
#include <stdio.h>

@interface MyNode:NSObject
@property int data;
@property(retain) MyNode *next;
@end

@implementation MyNode
@synthesize data, next;
@end


int main() {
  printf("Creating list\n");
  MyNode *head = [[MyNode alloc] init];

  for (int i = 0; i < 100; i++) {
    MyNode *next = [[MyNode alloc] init];
    [next setNext:[head next]];
    [next setData:i];
    [head setNext:next];
  }

  printf("Summing list\n");
  int sum = 0;
  uint64_t start = mach_absolute_time();
  MyNode *ptr = head;
  for (int i = 0; i < 100000; i++) {
    ptr = head;
    while (ptr) {
      sum += [ptr data];
      ptr = [ptr next];
    }
  }
  uint64_t delta = mach_absolute_time() - start;
  printf("%f ns\n", (double)delta);

  printf("sum = %d\n", sum);
  return 0;
}

File: utils/benchmark/LinkedList/link.swift
@asmname("mach_absolute_time") func __mach_absolute_time__() -> UInt64

@final class Node {
  var next: Node?
  var data: Int

  init(n: Node?, d: Int) {
    next = n
    data = d
  }
}

print("Creating list\n")
var head = Node(nil, 0)
for i in 0..100 {
  head = Node(head, i)
}

let start = __mach_absolute_time__()
print("Summing list\n")
var sum = 0
var ptr = head
for i in 0..100000 {
  ptr = head
  while let nxt = ptr.next {
    sum += ptr.data
    ptr = nxt
  }
}

let delta = __mach_absolute_time__() - start
print("sum = \(sum)\n")
print("\(delta) nanoseconds.")
